onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc

```diff
@@ -0,0 +1,613 @@
+                     [](int64_t i) { return static_cast<uint32_t>(i); });
+    }
+    std::vector<uint32_t> new_new_a_shape, new_new_b_shape;
+    uint32_t a_dim = 1, b_dim = 1;
+    for (auto idx : a_1) {
+      new_new_a_shape.push_back(new_a_shape[idx]);
+    }
+    for (auto idx : a_2) {
+      new_new_a_shape.push_back(new_a_shape[idx]);
+    }
+    for (auto idx : a_3) {
+      a_dim *= new_a_shape[idx];
+    }
+    new_new_a_shape.push_back(a_dim);
+    for (auto idx : b_1) {
+      new_new_b_shape.push_back(new_b_shape[idx]);
+    }
+    for (auto idx : b_2) {
+      b_dim *= new_b_shape[idx];
+    }
+    new_new_b_shape.push_back(b_dim);
+    for (auto idx : b_3) {
+      new_new_b_shape.push_back(new_b_shape[idx]);
+    }
+
+    input_a = model_builder.GetBuilder().call<emscripten::val>("reshape", input_a,
+                                                               emscripten::val::array(new_new_a_shape));
+    input_b = model_builder.GetBuilder().call<emscripten::val>("reshape", input_b,
+                                                               emscripten::val::array(new_b_shape));
+  }
+
+  // Step 2. Matmul
+  output = model_builder.GetBuilder().call<emscripten::val>("matmul", input_a, input_b);
+  std::vector<uint32_t> output_indices = a_1;
+  output_indices.push_back(a_2.back());
+  output_indices.push_back(b_3.back());
+
+  /*
+    Step 3. Output Transpose:
+    Use the following fast permutation calculation algorithm
+    to calculate the permutation of transpose.
+    sequence X[] -> sequence Y[] : permutation P[]
+    X[S[i]] = i, Y[T[i]] = i, P[T[i]] = S[i]
+    output_indices is X and target_output_indices is Y
+  */
+  std::vector<uint32_t> target_output_indices(output_labels.begin(), output_labels.end());
+
+  // map output dim labels to 0 ~ n-1
+  std::vector<uint32_t> arr(output_indices.begin(), output_indices.end());
+  std::map<uint32_t, uint32_t> mapping;
+  std::sort(arr.begin(), arr.end());
+  for (size_t i = 0; i < arr.size(); i++) {
+    mapping[arr[i]] = i;
+  }
+
+  for (size_t i = 0; i < output_indices.size(); i++) {
+    output_indices[i] = mapping[output_indices[i]];
+    if (i < target_output_indices.size()) {
+      target_output_indices[i] = mapping[target_output_indices[i]];
+    }
+  }
+
+  uint32_t p = target_output_indices.size();
+  std::vector<int64_t> s(output_indices.size(), -1), t(output_indices.size(), -1);
+  std::vector<uint32_t> v(output_indices.size(), 0);
+  for (uint32_t i = 0; i < output_indices.size(); ++i) {
+    s[output_indices[i]] = i;
+    if (i < target_output_indices.size()) {
+      t[target_output_indices[i]] = i;
+    }
+  }
+  for (uint32_t i = 0; i < output_indices.size(); ++i) {
+    if (t[i] == -1) {
+      t[i] = p++;
+    }
+    v[static_cast<uint32_t>(t[i])] = static_cast<uint32_t>(s[i]);
+  }
+
+  std::vector<uint32_t> sequence_o(output_indices.size());
+  std::iota(sequence_o.begin(), sequence_o.end(), 0);
+  if (v != sequence_o) {
```

Comment:
```
  if (output_permutation != sequence_o) {
```
We should use a clearer name than just "v". With the common exception of `i` for loop counters and `x`,`y`,`z` for coordinates, single letter variable names are dubious. Same for `s` and `t`.